From 4b6373977a0062688585489d7e8c60995977c3af Mon Sep 17 00:00:00 2001
From: James Calligeros <jcalligeros99@gmail.com>
Date: Sat, 1 Mar 2025 16:22:55 +1000
Subject: [PATCH 1/2] Support not running as root

We don't need to be running as root. Let's run as speakersafetyd
by default.

Signed-off-by: James Calligeros <jcalligeros99@gmail.com>
---
 Makefile                | 5 ++++-
 README.md               | 1 +
 speakersafetyd.service  | 3 +++
 speakersafetyd.tmpfiles | 3 ++-
 src/main.rs             | 2 +-
 5 files changed, 11 insertions(+), 3 deletions(-)

diff --git a/Makefile b/Makefile
index b7224d0..665cca1 100644
--- a/Makefile
+++ b/Makefile
@@ -7,6 +7,8 @@ UDEVDIR ?= /lib/udev/rules.d
 TMPFILESDIR ?= /usr/lib/tmpfiles.d
 SHAREDIR ?= /usr/share/
 VARDIR ?= /var/
+SPEAKERSAFETYD_GROUP ?= speakersafetyd
+SPEAKERSAFETYD_USER ?= speakersafetyd
 
 all:
 	cargo build --release
@@ -22,9 +24,10 @@ install-data:
 	install -pm0644 95-speakersafetyd.rules $(DESTDIR)/$(UDEVDIR)/95-speakersafetyd.rules
 	install -dDm0755 $(DESTDIR)/$(SHAREDIR)/speakersafetyd/apple
 	install -pm0644 -t $(DESTDIR)/$(SHAREDIR)/speakersafetyd/apple $(wildcard conf/apple/*)
-	install -dDm0755 $(DESTDIR)/$(VARDIR)/lib/speakersafetyd/blackbox
+	install -dDm0755 -o $(SPEAKERSAFETYD_USER) -g $(SPEAKERSAFETYD_GROUP) $(DESTDIR)/$(VARDIR)/lib/speakersafetyd/blackbox
 	install -dDm0755 $(DESTDIR)/$(TMPFILESDIR)
 	install -pm0644 speakersafetyd.tmpfiles $(DESTDIR)/$(TMPFILESDIR)/speakersafetyd.conf
+	install -dDm0755 -o $(SPEAKERSAFETYD_USER) -g $(SPEAKERSAFETYD_GROUP) /run/speakersafetyd
 
 uninstall:
 	rm -f $(DESTDIR)/$(BINDIR)/speakersafetyd $(DESTDIR)/$(UNITDIR)/speakersafetyd.service $(DESTDIR)/$(UDEVDIR)/95-speakersafetyd.rules $(DESTDIR)/$(TMPFILESDIR)/speakersafetyd.conf
diff --git a/README.md b/README.md
index b256625..1970d43 100644
--- a/README.md
+++ b/README.md
@@ -25,6 +25,7 @@ adapt for any device that provides V/ISENSE data in a manner similar to TAS2764.
 * Rust stable
 * alsa-lib
 * An Apple Silicon Mac running Asahi Linux
+* A `speakersafetyd` user in the `audio` group
 
 ### Some background on Smart Amps
 The cheap component speaker elements used in modern devices like
diff --git a/speakersafetyd.service b/speakersafetyd.service
index b7c30ad..a1bfe22 100644
--- a/speakersafetyd.service
+++ b/speakersafetyd.service
@@ -4,6 +4,9 @@ Description=Speaker Protection Daemon
 [Service]
 Type=simple
 ExecStart=/usr/bin/speakersafetyd -c /usr/share/speakersafetyd/ -b /var/lib/speakersafetyd/blackbox -m 7
+User=speakersafetyd
+AmbientCapabilities=CAP_SYS_NICE
+CapabilityBoundingSet=CAP_SYS_NICE
 UMask=0066
 Restart=on-failure
 RestartSec=1
diff --git a/speakersafetyd.tmpfiles b/speakersafetyd.tmpfiles
index 62a9560..7d81efe 100644
--- a/speakersafetyd.tmpfiles
+++ b/speakersafetyd.tmpfiles
@@ -1 +1,2 @@
-d /var/lib/speakersafetyd/blackbox 0755 root root -
+d /var/lib/speakersafetyd/blackbox 0755 speakersafetyd speakersafetyd -
+d /run/speakersafetyd 0755 speakersafetyd speakersafetyd -
diff --git a/src/main.rs b/src/main.rs
index 8564d4b..85ba486 100644
--- a/src/main.rs
+++ b/src/main.rs
@@ -30,7 +30,7 @@ const DEFAULT_CONFIG_PATH: &str = "share/speakersafetyd";
 
 const UNLOCK_MAGIC: i32 = 0xdec1be15u32 as i32;
 
-const FLAGFILE: &str = "/run/speakersafetyd.flag";
+const FLAGFILE: &str = "/run/speakersafetyd/speakersafetyd.flag";
 
 /// Simple program to greet a person
 #[derive(Parser, Debug)]
-- 
2.47.2

